### PR TITLE
feat(bot): add telegram interaction

### PR DESCRIPTION
**Pull request by romutchio**
Implementation of 2 commands:
* /update -- calls keyboard with filters and apply them on advising
* /search -- full text search on wines

### CODE REVIEW

File: src/main/scala/bot/domain/AdviseState.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  priceMin: Option[Int],
+  priceMax: Option[Int],
+) {
+  def toCountryCode: Option[CountryCode] = this.country.flatMap(AdviseState.CountryButtonToCountryCode.get)
+  def toWineType: Option[WineType] = this.wineType.flatMap(AdviseState.WineTypeButtonToWineType.get)
+}
+
+object AdviseState {
+  private val CountryButtonToCountryCode: Map[String, CountryCode] = Map(
+    (FranceCountryButton.name,       France),
```
Comment (romutchio) [inline src/main/scala/bot/domain/AdviseState.scala]: How to get a guarantee on a type level that all CountryButtons are mentioned here?

File: src/main/scala/bot/MessageFormatter.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  def adviseStateToMessageNotFound(adviseState: AdviseState): String = {
+    List(
+      Some("*Nothing was found* for selected options:"),
+      Some(""),
+      adviseState.country.map(x => s"*Country:* $x"),
+      adviseState.wineType.map(x => s"*Type:* $x"),
+      adviseState.priceMin.map(x => s"*Price from:* € $x"),
+      adviseState.priceMax.map(x => s"*Price to:* € $x")
+    ).flatten.mkString("\n")
```
Comment (ki4kis) [inline src/main/scala/bot/MessageFormatter.scala]: It's not really nice to build a message like this.
Not easily understandable what the end result is supposed to be.

File: src/main/scala/bot/MessageFormatter.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      Some("Advice options:"),
+      Some(""),
+      adviseState.country.map(x => s"*Country:* $x"),
+      adviseState.wineType.map(x => s"*Type:* $x"),
+      adviseState.priceMin.map(x => s"*Price from:* € $x"),
+      adviseState.priceMax.map(x => s"*Price to:* € $x")
+    ).flatten.mkString("\n")
+  }
+
+  def adviseStateToMessageNotFound(adviseState: AdviseState): String = {
```
Comment (ki4kis) [inline src/main/scala/bot/MessageFormatter.scala]: No need to name function as `<something>To<something else>`, we can figure this out from the function signature 😄

File: src/main/scala/bot/PerChatState.scala [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package bot
+
+import bot.domain.AdviseState
+import com.bot4s.telegram.models.{CallbackQuery, Message}
+
+
+trait PerChatState[F[_]] {
+  private val chatState = collection.mutable.Map[Long, AdviseState]()
```
Comment (ki4kis) [inline src/main/scala/bot/PerChatState.scala]: Replace the `mutable.Map[Long, AdviseState]` with a cats `Ref[immutable.Map[Long, AdviseState]]`

File: src/main/scala/bot/domain/buttons/Button.scala [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  val tag: String
+
+  private def button: String = (emoji, name) match {
+    case (Some(e), n) => s"$e $n"
+    case (None, n) => n
+  }
+
+  private def prefixTag(tag: String)(s: String): String = tag + s
+
+  def toInlineKeyboardButton: InlineKeyboardButton =
```
Comment (ki4kis) [inline src/main/scala/bot/domain/buttons/Button.scala]: I am not sure you need this method to be so verbose

File: src/main/scala/bot/domain/buttons/ButtonMarkup.scala
Comment (ki4kis) [inline src/main/scala/bot/domain/buttons/ButtonMarkup.scala:1]: Optional, but I would like it, if the code would reflect the way the buttons are visualized
```scala
Seq(
  Seq(A, B, C),
  Seq(D, E),
  Seq(F, G, H)
)
```

File: src/main/scala/client/WineBot.scala [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            } yield ()
+        }
+      } yield ()
+
+      maybeEditFuture.getOrElse(Async[F].pure()).void
+    }
+  }
+
+  private def createCallback(
+    implicit cbq: CallbackQuery,
```
Comment (ki4kis) [inline src/main/scala/client/WineBot.scala]: implicits should be curried separately from the normal params. 
```scala
def function(normalParams...)(implicit implicitParams...)
```
If formatted like you have all other params have become implicits as well.

**Pull request merged**